packages/widgets/Source/Viewer/Viewer.js

```diff
@@ -680,10 +680,21 @@ Either specify options.terrainProvider instead or set options.baseLayerPicker to
     scene.imageryLayers.addImageryProvider(options.imageryProvider);
   }
   if (defined(options.terrainProvider)) {
-    if (createBaseLayerPicker) {
-      baseLayerPicker.viewModel.selectedTerrain = undefined;
+    // options.terrainProvider is a promise
+    // Promise.resolve is not used so that synchronous code can immediately execute
+    if (defined(options.terrainProvider.then)) {
+      options.terrainProvider.then((provider) => {
```

Comment:
I noticed code like this in a few places and it's still problematic. Is this temporary code for backwards compatibility or permanent code (ideally any temporary code would have a comment stating as such with a link back to the GH issue to remove it)

If it's temporary, it can stay as is., if it's permanent:

* What happens if the Viewer is destroyed before before this promise resolves?
* What happens if code inside this promise rejects for some reason?

Reply:
I just realized something. Why are you taking a promise for `terrainProvider` at all? Wouldn't it just be easier to expect the user to await the call and then removes this problem completely"? Async stuff in a constructor is almost always a bad call.

Reply:
This case was intended to be permanent. (Some of the `baseLayerPicker` and Cesium Viewer app changes could be a bit cleaner as I alluded to in the PR description, but I would like to address those once the imageryProvider deprecations are in as well.)

> What happens if the Viewer is destroyed before before this promise resolves?

The Viewer successfully destroys without error. Just in case, I will add an early return if `isDestroyed` is true.

> What happens if code inside this promise rejects for some reason?

For example, if the code is the following:

```
const viewer = new Cesium.Viewer("cesiumContainer", {
  terrainProvider: Cesium.createWorldTerrainAsync(),
});
```

and `createWorldTerrainAsync` throws, the result is the error is uncaught and shows in the console with the error stack. The terrain is not created, but the scene continues its render loop. For example in Sandcastle:

<img width="394" alt="image" src="https://user-images.githubusercontent.com/4439461/213496110-7c8e16cb-eb67-450d-962a-c401229aff06.png">

I thought the point of something like this was to ensure the errors remain uncaught such that the app can deal with it accordingly rather than fail silently. If we'd rather catch the error explicitly and log something to the console, we can do that instead.

Reply:
> I thought the point of something like this was to ensure the errors remain uncaught such that the app can deal with it

In Node, an uncaught promises crashes the app immediately, so there is no way to deal with it. Even in the above callstack, there's no way for our browser users to handle the promise. It will always be uncaught, right?

Either way, if you're up for removing the async nature completely, I think it's definitely worth a try and removes this issue. Ultimately it's just `terrainProvider: await createWorldTerrainAsync()` instead of `terrainProvider: createWorldTerrain()`, right? So it might not be that bad in most cases.